fleetspeak/src/server/internal/cache/cache.go

```diff
@@ -0,0 +1,114 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/google/fleetspeak/fleetspeak/src/common"
+	"github.com/google/fleetspeak/fleetspeak/src/server/db"
+)
+
+// Controls cache timing behavior, variable only to support unit tests.
+var (
+	maxAge         = 30 * time.Second
+	expireInterval = 5 * time.Minute
+)
+
+// Client is a cache of recently connected clients.
+type Client struct {
+	m    map[common.ClientID]*clientEntry
+	l    sync.RWMutex
+	stop chan struct{}
+}
+
+type clientEntry struct {
+	u time.Time
+	d *db.ClientData
+}
+
+// NewClient returns a new cache of client data.
+func NewClient() *Client {
+	ret := &Client{
+		m:    make(map[common.ClientID]*clientEntry),
+		stop: make(chan struct{}),
+	}
+	go ret.expireLoop()
+	return ret
+}
+
+// Get returns the cached client data, if there is sufficiently fresh data in
+// the cache, otherwise nil.
+func (c *Client) Get(id common.ClientID) *db.ClientData {
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	e := c.m[id]
+	if e == nil || db.Now().Sub(e.u) > maxAge {
+		return nil
+	}
+	return e.d
+}
+
+// Update updates or sets the cached data for a particular client. If data is
+// nil, it clears the data for the client.
+func (c *Client) Update(id common.ClientID, data *db.ClientData) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	if data == nil {
+		delete(c.m, id)
+	} else {
+		c.m[id] = &clientEntry{
+			u: db.Now(),
+			d: data,
+		}
+	}
+}
+
+// Clear empties the cache, removing all entries.
+func (c *Client) Clear() {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	c.m = make(map[common.ClientID]*clientEntry)
+}
+
+// Stop releases the resources required for background cache maintenence. The
+// Cache should not be used once Stop has been called.
+func (c *Client) Stop() {
+	close(c.stop)
+}
+
+// Size returns the current size taken up by the cache, this is a count of
+// client records, some of which may be too old to be visible.
+func (c *Client) Size() int {
+	c.l.RLock()
+	defer c.l.RUnlock()
+	return len(c.m)
+}
+
+func (c *Client) expireLoop() {
+	t := time.NewTicker(expireInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+			c.expire()
+		case <-c.stop:
+			return
+		}
+	}
+}
+
+// expire prunes the cache to clean out clients that are no longer up to date.
+func (c *Client) expire() {
```

Comment:
How big of a hiccup can this be worst case? 100ms? 500ms? Are we okay with having all callers of the cache wait this long? Just asking.

Reply:
I didn't think in terms of read/write locks previously. Having an option to read-lock helps a lot here - we can do one read-lock pass noting all the keys to be deleted, then acquire write-lock and delete the noted keys.

Reply:
I think this is unlikely to take even 100ms - how long do you think it takes to walk a 100k element has table? And even if it takes a bit, a hiccoup every 5 minutes is pretty minor. A two pass solution requires new allocations, is therefore less cache friendly, and subtle to get right - once you release the initial read lock, the records you plan to delete could be updated, so under the write lock you need to check again the expiry condition for the records you plan to remove.

Actually though, I just found https://github.com/golang/go/issues/20135. So the right answer may be to either do nothing, or to rebuild the map entirely. (of course there is a cost for allowing deletion during iteration)

Reply:
You should be able to promote a read lock to a write lock, no? Then the described race condition doesn't occur.

At any rate, not cleaning up sounds like the best way here. There are no bugs if there is no code & this structure will never get significantly large.

Reply:
Sane Mutex implementation normally just disallow lock promotion altogether as it makes deadlocks too easy. (suppose two threads try to promote at once)

Reply:
A memory leak is a bug, and the usual definition of a leak is that you hold on to data that you won't use. When leaks become a problem, they can be really hard to debug. Also, while deleting from the map might not shrink it, it still clears out the records so that the map space can be reused.